### ISSUE TITLE
Soft-fail certificate revocation check implementation

### ISSUE THREAD

**Issue by souch** [randombit/botan#1458]
HI,

I would like to have soft fail revocation check implemented in Botan. I don't think it can do that right now.
The best would be to have it on OCSP and CRL but right now only OCSP is plugged to the x509_path_validate. So I will try first to implement it with OCSP only.

But I don't know which way is best to implement it. And if you are ok with that. I will try to make a pull request once sthg ok has been done.
I first tried to add a parameter "revocation_softfail" to "Path_Validation_Restrictions" 

```
      * @param revocation_softfail if true, revocation check do not fail if
      *        revocation server is not available or revocation URL not set on
      *        certificate.
      */
      Path_Validation_Restrictions(bool require_rev = false,
                                   size_t minimum_key_strength = 110,
                                   bool ocsp_all_intermediates = false,
                                   bool revocation_softfail = false);
```

But I am a bit stuck right now on how to put that nicely in check_ocsp, check_ocsp_online,  Path_Validation_Result::successful_validation and x509_path_validate functions. I think I will need some new Certificate_Status_Code: NO_REVOCATION_URL and REVOCATION_SERVER_NOT_AVAILABLE.

I know this issue is a bit of a draft, but it is mainly to have your opinion and to have some guidelines and tips.

**Comment by randombit** [comment]
If I'm understanding what you want here, I don't think there is any reason to add an additional argument to Path_Validation_Restrictions. Instead just add the new status codes as warnings, and insert them for every cert where no revocation data was available (that happens in `PKIX::merge_revocation_status`). Applications which ignore the warning state of path validation results will work as usual, but an application which cares about revocation status can check.

**Comment by randombit** [comment]
Fixed with #1462?

**Comment by souch** [comment]
Yep. At least for OCSP. I don't need CRL right now.
